File: reference/aeonscript/decoder.py
"""High-level AeonScript decoder.

Inverse of encoder.py:
    list[oligo DNA] → concatenated bytes → RS-decoded → tag-stripped → payload
"""

from __future__ import annotations

from typing import Iterable

from .error_correction import decode_with_ecc
from .physical import dna_to_bytes
from .scrambler import descramble
from .semantic import parse_tag, validate_tag_dict

DEFAULT_RS_PARITY = 32
DEFAULT_RS_BLOCK = 223


def decode_oligos(
    oligos: Iterable[str],
    *,
    rs_parity: int = DEFAULT_RS_PARITY,
    rs_block_size: int = DEFAULT_RS_BLOCK,
    return_metadata: bool = False,
) -> bytes | tuple[bytes, dict]:
    """Decode a list of AeonScript DNA oligos back to the original payload.

    Parameters
    ----------
    oligos : iterable of DNA strings
    rs_parity, rs_block_size : Reed-Solomon parameters (must match encoder)
    return_metadata : if True, also return the parsed semantic tag dict and
                      number of corrected errors.

    Returns
    -------
    bytes
        The decoded payload (excluding the semantic tag).
    Or, if return_metadata=True:
    tuple[bytes, dict]
        (payload, {"tag": {...}, "errors_corrected": int})
    """
    # ---- L1: oligos → bytes --------------------------------------------
    rs_encoded = bytearray()
    for oligo in oligos:
        rs_encoded.extend(dna_to_bytes(oligo))

    # ---- Descramble (inverse of encoder PRBS) -------------------------
    descrambled = descramble(bytes(rs_encoded))

    # ---- L4: Reed-Solomon decode IF the stream looks RS-encoded -------
    # An RS-encoded stream has length = N × (block_size + nsym) = N × 255.
    # Otherwise the stream is plaintext (L4=none) — the canonical L1+L5
    # test vectors at spec/test-vectors/vectors-l1-l5.json use this path.
    full_block = rs_block_size + rs_parity
    if len(descrambled) >= full_block and len(descrambled) % full_block == 0:
        tagged_data, n_errors = decode_with_ecc(
            descrambled, nsym=rs_parity, block_size=rs_block_size
        )
    else:
        tagged_data, n_errors = descrambled, 0

    # ---- L5: parse and strip semantic tag ------------------------------
    # The tag starts with the byte '|' (0x7C) and ends at the next '|'.
    if not tagged_data.startswith(b"|"):
        raise ValueError("Decoded stream does not start with a AeonScript tag '|'")
    end_idx = tagged_data.index(b"|", 1)
    tag_str = tagged_data[: end_idx + 1].decode("ascii")
    payload_start = end_idx + 1

    tag_dict = parse_tag(tag_str)
    ok, msg = validate_tag_dict(tag_dict)
    if not ok:
        raise ValueError(f"Invalid semantic tag: {msg}")

    declared_len = int(tag_dict["LEN"])
    raw_payload = tagged_data[payload_start:]

    # The RS encoder may have padded the last block with zeros; trust LEN.
    payload = raw_payload[:declared_len]

    if return_metadata:
        return payload, {"tag": tag_dict, "errors_corrected": n_errors}
    return payload

File: reference-rs/src/decoder.rs
//! High-level AeonScript decoder.
//!
//! Inverse of [`crate::encoder`]:
//!
//! 1. L1 — concatenate decoded bytes from each oligo.
//! 2. Descrambler — XOR with the same PRBS stream (self-inverse).
//! 3. L4 — Reed-Solomon decode (skeleton in v0.1, passes through).
//! 4. L5 — parse and strip the semantic tag.
//! 5. Trim the payload to the declared length.

use crate::error_correction::{decode_with_ecc, EccError, DEFAULT_BLOCK, DEFAULT_NSYM};
use crate::physical::{dna_to_bytes, PhysicalError};
use crate::scrambler::descramble;
use crate::semantic::{parse_tag, validate_tag, TagError};

/// Errors produced by the decoder.
#[derive(Debug, thiserror::Error)]
#[allow(missing_docs)]
pub enum DecodeError {
    #[error("L1 decode failed: {0}")]
    Physical(#[from] PhysicalError),
    #[error("L4 decode failed: {0}")]
    Ecc(#[from] EccError),
    #[error("L5 tag error: {0}")]
    Tag(#[from] TagError),
    #[error("Decoded stream does not start with a tag '|'")]
    MissingTagStart,
    #[error("Decoded stream is missing the closing '|' of the tag")]
    MissingTagEnd,
    #[error("Tag is not valid ASCII")]
    NonAsciiTag,
}

/// Decode a list of AeonScript DNA oligos back to the original payload bytes.
pub fn decode_oligos(oligos: &[String]) -> Result<Vec<u8>, DecodeError> {
    // L1: oligos → bytes
    let mut rs_encoded = Vec::new();
    for o in oligos {
        rs_encoded.extend(dna_to_bytes(o)?);
    }

    // Descrambler
    let plain = descramble(&rs_encoded);

    // L4: Reed-Solomon if the byte stream looks RS-encoded (length is a
    // positive multiple of the full block size 255). Otherwise we assume
    // L4=none (plaintext) — which is what the canonical L1+L5 test vectors
    // use.
    let full_block = DEFAULT_BLOCK + DEFAULT_NSYM;
    let after_l4 = if plain.len() >= full_block && plain.len() % full_block == 0 {
        match decode_with_ecc(&plain, DEFAULT_NSYM, DEFAULT_BLOCK) {
            Ok((decoded, _)) => decoded,
            // RS decode failure on aligned input is genuine corruption.
            Err(e) => return Err(DecodeError::Ecc(e)),
        }
    } else {
        plain
    };

    // L5: find the tag
    if after_l4.first() != Some(&b'|') {
        return Err(DecodeError::MissingTagStart);
    }
    let end_idx = after_l4
        .iter()
        .skip(1)
        .position(|&b| b == b'|')
        .map(|p| p + 1)
        .ok_or(DecodeError::MissingTagEnd)?;

    let tag_bytes = &after_l4[..=end_idx];
    let tag_str = std::str::from_utf8(tag_bytes).map_err(|_| DecodeError::NonAsciiTag)?;
    let tag = parse_tag(tag_str)?;
    validate_tag(&tag)?;

    let len = tag.payload_len().unwrap_or(0);
    let payload_start = end_idx + 1;
    let payload_end = (payload_start + len).min(after_l4.len());
    Ok(after_l4[payload_start..payload_end].to_vec())
}
